ml/src/main/scala/ai/h2o/sparkling/ml/models/H2OTargetEncoderMojoModel.scala

```diff
@@ -0,0 +1,84 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package ai.h2o.sparkling.ml.models
+
+
+import ai.h2o.sparkling.ml.features.H2OTargetEncoderBase
+
+import hex.genmodel.easy.EasyPredictModelWrapper
+import hex.genmodel.algos.targetencoder.TargetEncoderMojoModel
+
+import org.apache.spark.h2o.converters.RowConverter
+import org.apache.spark.ml.Model
+import org.apache.spark.ml.h2o.models.{H2OMOJOFlattenedInput, H2OMOJOReadable, H2OMOJOWritable}
+import org.apache.spark.ml.param.ParamMap
+import org.apache.spark.ml.util.Identifiable
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.{DataFrame, Dataset, Row}
+import org.apache.spark.sql.DatasetExtensions._
+
+import water.support.ModelSerializationSupport
+
+class H2OTargetEncoderMojoModel(override val uid: String) extends Model[H2OTargetEncoderMojoModel]
+  with H2OTargetEncoderBase with H2OMOJOWritable with H2OMOJOFlattenedInput {
+
+  override protected def inputColumnNames: Array[String] = getInputCols()
+
+  override protected def outputColumnName: String = this.getClass.getSimpleName + "_output"
+
+  def this() = this(Identifiable.randomUID("H2OTargetEncoderMojoModel"))
+
+  override def transform(dataset: Dataset[_]): DataFrame = {
+    val outputCols = getOutputCols()
+    val udfWrapper = H2OTargetEncoderMojoUdfWrapper(getMojoData(), outputCols)
+    val withPredictionsDF = applyPredictionUdf(dataset, _ => udfWrapper.mojoUdf)
+    withPredictionsDF
+      .withColumns(outputCols, outputCols.zipWithIndex.map{ case (c, i) => col(outputColumnName)(i) as c })
+      .drop(outputColumnName)
+  }
+
+  override def copy(extra: ParamMap): H2OTargetEncoderMojoModel = defaultCopy(extra)
+}
+
+/**
+  * The class holds all necessary dependencies of udf that needs to be serialized.
+  */
+case class H2OTargetEncoderMojoUdfWrapper(mojoData: Array[Byte], outputCols: Array[String]) {
+  @transient private lazy val easyPredictModelWrapper: EasyPredictModelWrapper = {
+    val model = ModelSerializationSupport
+      .getMojoModel(mojoData)
+      .asInstanceOf[TargetEncoderMojoModel]
+    val config = new EasyPredictModelWrapper.Config()
+    config.setModel(model)
+    config.setConvertUnknownCategoricalLevelsToNa(true)
```

Comment:
It this hardcoded in H2O as well? Should this be configurable via our H2OMojoSettings?

Reply:
If this option was false, `EasyPredictModelWrapper` would throw an exception and [output would be null](https://github.com/h2oai/sparkling-water/pull/1192/files#diff-e51c0ac92a4dbe5601ffae48ae3dfb4fR79). In a regular case, TE uses a global average as output value for unknown values.

I can expose this option, but is there use case for that?

Reply:
No need to expose, thanks for explanation!

Reply:
Let's double-check it with @deil87 and @michalkurka.

Can you think of an use case where would make more sense if an algorithm in a pipeline after TE would obtain NA (null -> NA) instead of the global average for unknown categoricals?

Btw: `TargetEncoderModel` (Java API) throws an exception in the same scenario:
```
water.exceptions.H2OConcurrentModificationException: Rollups not possible, because Vec was deleted: $04ff5e000000ffffffffb00376b65fbfebb9ad4aabb7d387d3$D

	at water.fvec.RollupStats.get(RollupStats.java:323)
	at water.fvec.RollupStats.get(RollupStats.java:353)
	at water.fvec.Vec.rollupStats(Vec.java:901)
	at water.fvec.Vec.naCnt(Vec.java:853)
	at ai.h2o.automl.targetencoding.TargetEncoder.imputeWithMean(TargetEncoder.java:294)
	at ai.h2o.automl.targetencoding.TargetEncoder.applyTargetEncoding(TargetEncoder.java:645)
	at ai.h2o.automl.targetencoding.TargetEncoderModel.transform(TargetEncoderModel.java:96)
	at ai.h2o.sparkling.ml.models.H2OTargetEncoderModel.transformTrainingDataset(H2OTargetEncoderModel.scala:61)
	at ai.h2o.sparkling.ml.features.H2OTargetEncoderTestSuite$$anonfun$6.apply$mcV$sp(H2OTargetEncoderTestSuite.scala:137)
	at ai.h2o.sparkling.ml.features.H2OTargetEncoderTestSuite$$anonfun$6.apply(H2OTargetEncoderTestSuite.scala:125)
	at ai.h2o.sparkling.ml.features.H2OTargetEncoderTestSuite$$anonfun$6.apply(H2OTargetEncoderTestSuite.scala:125)
	at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
```
  
But this is not a Sparkling Water use case.

Reply:
@mn-mikke This is hardcoded in TargetEncoder because along the way I have found a [bug](https://0xdata.atlassian.net/browse/PUBDEV-5924) in merge functionality.... and I had to always substitute nulls with new category. We probably can remove this parameter at all later.

Reply:
https://0xdata.atlassian.net/browse/PUBDEV-6704